### PR TITLE
[WIP]Add checkpoint support for the secondary instance and ldb

### PR DESCRIPTION
**Pull request by gitbw95**
### Summary
ldb cannot checkpoint when using with -secondary_path.

CreateCheckPoint may call GetLiveFilesStorageInfo that may flush memtable. Secondary instance does not trigger flush, so the flush never returns. Thus,  implement a GetLiveFilesStorageInfo() for DBImplSecondary.

### Test Plans
Following steps should successfully checkpoint the secondary instance:
ldb put foo bar --db=/mydb --create_if_missing
ldb put foo2 bar2 --db=/mydb
ldb get foo --db=/mydb --secondary_path=/my_sec_path
ldb checkpoint --db=/mydb --secondary_path=/my_sec_path --checkpoint_dir=/checkpoint_db

### CODE REVIEW

**Comment by gitbw95** [comment]
This PR replaces https://github.com/facebook/rocksdb/pull/9755

@pdillinger and @riversand963 , pls review.
Specially, do we want a GetLiveFilesStorageInfoHelper with the extra parameter is_flush_supported referring both to memtable flush and WAL flush. Note that "flush" is not a logical write, so could (at least in my mind) be supported on a "read only" DB interface. so that the GetLiveFilesStorageInfo functions in db_filesnapshot.cc and db_impl_secondary.cc can be simpler.


@pdillinger , pls advise the tests that I should add.

File: db/db_impl/db_impl_secondary.cc [hunk 1]
```diff
@@ -620,6 +620,44 @@ Status DBImplSecondary::TryCatchUpWithPrimary() {
...
+  files->clear();
+  std::vector<LiveFileStorageInfo> results;
+  Status s;
+
+  ROCKS_LOG_INFO(immutable_db_options_.info_log,
+                 "DBImplSecondary::GetLiveFilesStorageInfo");
+  LogFlush(immutable_db_options_.info_log);
+
+  bool need_flush_memtable{true};
+  s = IsFlushMemtableNeededForGetLiveFiles(opts, need_flush_memtable);
```
Comment (riversand963) [inline db/db_impl/db_impl_secondary.cc:636]: For secondary instance, `need_flush_memtable` should always be false. Why call a function to evaluate it?

File: db/db_impl/db_impl_secondary.h [hunk 2]
```diff
@@ -172,6 +172,12 @@ class DBImplSecondary : public DBImpl {
     return Status::NotSupported("Not supported operation in secondary mode.");
   }
 
+  // For the secondary instance, this function only get live files
+  // and doesn't flush memtable.
+  virtual Status GetLiveFilesStorageInfo(
```
Comment (riversand963) [inline db/db_impl/db_impl_secondary.h:177]: Nit: remove `virtual`

File: db/db_filesnapshot.cc [hunk 3]
```diff
@@ -195,24 +187,20 @@ Status DBImpl::GetLiveFilesStorageInfo(
...
-        flush_memtable = false;
+        need_flush_memtable = false;
       }
       live_wal_files.clear();
     }
   }
+  return s;
+}
 
+void DBImpl::GetNonWALLiveFiles(const LiveFilesStorageInfoOptions& opts,
```
Comment (riversand963) [inline db/db_filesnapshot.cc:198]: Nit: I feel `GetLiveFilesExceptWals` is a better name.

File: db/db_filesnapshot.cc [hunk 4]
```diff
@@ -341,24 +329,16 @@ Status DBImpl::GetLiveFilesStorageInfo(
...
-  TEST_SYNC_POINT("CheckpointImpl::CreateCheckpoint:SavedLiveFiles2");
-
-  if (s.ok()) {
-    s = FlushWAL(false /* sync */);
-  }
-
-  TEST_SYNC_POINT("CheckpointImpl::CreateCustomCheckpoint:AfterGetLive1");
-  TEST_SYNC_POINT("CheckpointImpl::CreateCustomCheckpoint:AfterGetLive2");
-
+Status DBImpl::GetWALLiveFiles(const LiveFilesStorageInfoOptions& opts,
```
Comment (riversand963) [inline db/db_filesnapshot.cc:334]: Nit: `GetLiveWals()` sounds better?

File: db/db_filesnapshot.cc [hunk 5]
```diff
@@ -163,22 +163,14 @@ Status DBImpl::GetCurrentWalFile(std::unique_ptr<LogFile>* current_log_file) {
...
-    std::vector<LiveFileStorageInfo>* files) {
-  // To avoid returning partial results, only move results to files on success.
-  assert(files);
-  files->clear();
-  std::vector<LiveFileStorageInfo> results;
-
-  // NOTE: This implementation was largely migrated from Checkpoint.
-
+Status DBImpl::IsFlushMemtableNeededForGetLiveFiles(
+    const LiveFilesStorageInfoOptions& opts, bool& need_flush_memtable) {
```
Comment (riversand963) [inline db/db_filesnapshot.cc:167]: Not very important, but I think returning a <Status, bool> seems clearer for people reading the code.

File: db/db_filesnapshot.cc (hunk 3 above)
Comment (riversand963) [reply db/db_filesnapshot.cc:198]: The function name suggests that it is about non-wal files, but `min_log_num` is an output argument. Can we remove it and put it out of this function? Can we evaluate it in `GetLiveWals()` below?

Comment (gitbw95) [reply db/db_filesnapshot.cc:198]: In the original GetLiveFilesStorageInfo() , min_log_num is obtained within a mutex lock. I am not sure whether it is fine to get it later in the GetLiveWals().

File: db/db_impl/db_impl_secondary.cc (hunk 1 above)
Comment (gitbw95) [reply db/db_impl/db_impl_secondary.cc:636]: Will remove.

File: db/db_filesnapshot.cc (hunk 4 above)
Comment (gitbw95) [reply db/db_filesnapshot.cc:334]: Will change.

File: db/db_filesnapshot.cc (hunk 5 above)
Comment (gitbw95) [reply db/db_filesnapshot.cc:167]: Will keep as is.

File: db/db_impl/db_impl.h [hunk 6]
```diff
@@ -1281,6 +1281,17 @@ class DBImpl : public DB {
       ColumnFamilyData* cfd, const ExternalSstFileIngestionJob& ingestion_job);
 
   Status FlushForGetLiveFiles();
+
+  Status IsFlushMemtableNeededForGetLiveFiles(
+      const LiveFilesStorageInfoOptions& opts, bool& need_flush_memtable);
+  void GetLiveFilesExceptWals(const LiveFilesStorageInfoOptions& opts,
+                              std::vector<LiveFileStorageInfo>& results,
```
Comment (pdillinger) [inline db/db_impl/db_impl.h:1288]: I, and IIRC the rest of the team with opinions, prefer the old Google style that did not allow non-const reference parameters. Why? Readability. (I shouldn't have to know the signatures of called functions to construct reasonably accurate UD/DU chains on local variables - https://en.wikipedia.org/wiki/Use-define_chain)

**Review by pdillinger — COMMENTED** [review]
My opinion on testing:

Add secondary DB as an alternative parameterization (or similar) to an existing "easy" test such as CheckpointReadOnlyDBWithMultipleColumnFamilies. We might want to also add it to an essential "hard" test such as BackupEngineTest::ChangeManifestDuringBackupCreation also. (With Checkpoint support, Backup support should be automatic.)

And for stress/crash test: it would make sense to me that each of TestCheckpoint and TestBackupRestore would with some probability (1 in 3?) open the db dir as secondary to get the checkpoint/backup rather than using the already-open DB.

File: db/db_impl/db_impl.h (hunk 6 above)
Comment (pdillinger) [reply db/db_impl/db_impl.h:1288]: Actually, team may not agree, but here are some references about old style. Not sure why it changed.

https://stackoverflow.com/questions/26441220/googles-style-guide-about-input-output-parameters-as-pointers
https://www.youtube.com/watch?v=NOCElcMcFik (Seek to 15:45 and to 28:28)
https://github.com/google/styleguide/commit/7a7a2f510efe7d7fc#diff-bcadcf8be931ffdd5d6a65c60c266039cf1f96b7f35bfb772662db811214c5a0R1713

**Pull request closed without merging**